src/plugins/instrument_fetch.js

```diff
@@ -0,0 +1,269 @@
+    return result;
+}
+
+// Automatically create spans for all XMLHttpRequest objects.
+//
+// NOTE: this code currently works only with a single Tracer.
+//
+class InstrumentFetch {
+    constructor() {
+        this._enabled = this._isValidContext();
+        this._proxyInited = false;
+        this._internalExclusions = [];
+        this._tracer = null;
+        this._handleOptions = this._handleOptions.bind(this);
+
+        if (!this._enabled) {
+            return;
+        }
+    }
+
+    name() {
+        return 'instrument_fetch';
+    }
+
+    addOptions(tracerImp) {
+        tracerImp.addOption('fetch_instrumentation', { type : 'bool', defaultValue : false });
+        tracerImp.addOption('fetch_url_inclusion_patterns', { type : 'array', defaultValue : [/.*/] });
+        tracerImp.addOption('fetch_url_exclusion_patterns', { type : 'array', defaultValue : [] });
+    }
+
+    start(tracerImp) {
+        if (!this._enabled) {
+            return;
+        }
+        this._tracer = tracerImp;
+
+        let currentOptions = tracerImp.options();
+        this._addServiceHostToExclusions(currentOptions);
+        this._handleOptions({}, currentOptions);
+        tracerImp.on('options', this._handleOptions);
+    }
+
+    stop() {
+        if (!this._enabled) {
+            return;
+        }
+        window.fetch = proxiedFetch;
+    }
+
+    /**
+     * Respond to options changes on the Tracer.
+     *
+     * Note that `modified` is the options that have changed in this call,
+     * along with their previous and new values. `current` is the full set of
+     * current options *including* the newly modified values.
+     */
+    _handleOptions(modified, current) {
+        // Automatically add the service host itself to the list of exclusions
+        // to avoid reporting on the reports themselves
+        let serviceHost = modified.collector_host;
+        if (serviceHost) {
+            this._addServiceHostToExclusions(current);
+        }
+
+        // Set up the proxied XHR calls unless disabled
+        if (!this._proxyInited && current.fetch_instrumentation) {
+            this._proxyInited = true;
+            window.fetch = this._instrumentFetch();
+        }
+    }
+
+    /**
+     * Ensure that the reports to the collector don't get instrumented as well,
+     * as that recursive instrumentation is more confusing than valuable!
+     */
+    _addServiceHostToExclusions(opts) {
+        if (opts.collector_host.length === 0) {
+            return;
+        }
+
+        // http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex
+        function escapeRegExp(str) {
+            // return (`${str}`).replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
+        }
+
+        // Check against the hostname without the port as well as the canonicalized
+        // URL may drop the standard port.
+        let host = escapeRegExp(opts.collector_host);
+        let port = escapeRegExp(opts.collector_port);
```

Comment:
It’s an edge case, but what happens if we’re in a situation where you’re on a non-standard port like 8080?

Reply:
we were using 8081 in our local development and things seemed to work well. I think that the `set` array would just have the single entry with the full host + port. this code is just [directly lifted from the instrument_xhr plugin](https://github.com/lightstep/lightstep-tracer-javascript/blob/master/src/plugins/instrument_xhr.js#L140).

Reply:
👍 cool!